### DISCUSSION TITLE
Caching requests/responses?

### DISCUSSION THREAD

**Discussion by trifle** [synchronizing/mitm#10] [Q&A]
Hi, thanks for this wonderful library!
It's rare to see such compact and elegant async code.

I've been wondering whether it would be possible to add caching to mitm.
Middlewares would not work for this, since don't have the capability of fundamentally altering the data flow.

So my question is: Do you see an elegant way of adding such functionality?

The basic issue is that we currently cannot initiate a (fake/cached) server response when handling the request.

One possible option would be to check and prevent sending data in https://github.com/synchronizing/mitm/blob/master/mitm/mitm.py#L137.

I.e. something like

```python
if data:
    writer.write(data)
```

Thus the request would never get sent to the true destination if a middleware blanks out the request data.
A second middleware would then need to figure out how and when to return the cached data.

But that's of course not very robust and might break on all sorts of edge cases.

Perhaps you have a better idea?

**Comment by synchronizing** [comment]
Hey @trifle, thanks for the kind words regarding the project. 

Regarding caching, I don't think middlewares or protocols would be useful for this. The implementation of `mitm` would need to change in itself to allow this functionality. That's not to say this is a bad idea or use-case (I can def. see where this is useful), but it is to say this is currently not possible in this iteration. It _can_ be added, however. 

> The basic issue is that we currently cannot initiate a (fake/cached) server response when handling the request.

Could you expand a bit on this end? The typical solution to this problem is caching on the client side of things, but I'm assuming this is not feasible?

Also, check out [`mitmproxy`](https://mitmproxy.org/). Their codebase is a bit more complex (a lot more complex), but they seem to support caching requests/responses out of the box. See [here](https://github.com/mitmproxy/mitmproxy/issues/648).

**Reply by trifle** [reply]
Thanks for the quick reply!
My use case is rather simple: I'd like to cache outgoing requests from a number of internal clients to make sure we don't overload APIs or run into rate limits. So what I'm looking for is essentially a caching proxy like squid.

But on top of caching, I'd like to add some custom logic - for per-domain TTLs for example.

mitm seemed like a good starting point due to its compact and readable codebase.

I agree that it might be a bit of work. In essence, I'd need to parse http requests to cache based on query parameters and so on.
Next, I'd need to track request and response pairs to switch to cached responses on the client connection:

- new request -> send to original server and cache response
- known request -> load cached response and skip original server

I guess I'll poke this idea a bit and see how realistic it is.

**Reply by synchronizing** [reply]
I would need to think a little bit more about this problem to see what can be done as it relates to `mitm`. I began writing a response of something that could be done...

> > I'd like to cache outgoing requests from a number of internal clients to make sure we don't overload APIs or run into rate limits.
> 
> I see - this is definitely a very specific & unique problem set. 
> 
> One thing that might help is to split up the entire flow of the `MITM.mitm` function into separate functionalities:
> 
> https://github.com/synchronizing/mitm/blob/9fabb564a7bb6f3e6da42ba325f3d797aec6c9e4/mitm/mitm.py#L95
> 
> ```python
> class MITM(CoroutineClass):
>     def __init__(...)
>     def entry(...)
>     def client_connected(...) # Would call 'identify_protocol'.
>     def identify_protocol(...) # Would call 'connect_to_server'.
>     def connect_to_server(...) # Would call 'relay_messages'.
>     def relay_messages(...)
> ```
> 
> Allowing for custom implementations of the low-level `mitm` flow via inheritance (`CustomMITM(MITM)`). From the above, something like so could be implemented:
> 
> 1. When in `connect_to_server` `mitm` would check to see if the server has been previously connected to. 
>    a. If the server has been connected to we hold off connecting to it until we see what the client wants to relay in `relay_messages`. 
>    b. If the server has never been connected to do we do the usual logic (e.g. `super().connect_to_server(...)`).
> 2. When in `relay_messages`, `mitm` would:
>    a. If the sever has been connected to in the past (following (a) above), we get the last request being sent (say, a `GET`) and check it within our cache to see if we have seen it before. If we get a cache hit, we relay it to the client. If we don't get a cache hit, we call `connect_to_server` with some `force = True` flag (which will connect to server and call (b) below).
>    b. If the server has not been connected to in the past (following (b) above), we relay the messages, but cache the response.

... but what hit me after writing all of the above is that there might be cases where the flow of things is a little more complicated. That is, the client/server communicate back-and-forth via multiple messages (not a simple `CONNECT` -> `GET` -> response flow) as would be the case for something like OAuth, etc. Because of such cases a few things would need to be considered:

1. The `mitm` must cache req/resp flows - not just a singular request and a singular response. 
2. There would need to be a fail-safe mechanism that validates the req/resp flow as it happens. Perhaps the first 2 requests/responses are cacheable, but the third request/response is not - in such case, how should this be handled?

I'll revisit this question in the next coming days. For now, however, I'm unsure what we can do. I'm open to any ideas on modifying low-level `mitm` code as this project is/was intended to be used for larger projects just like this, and so this is a welcoming discussion.

**Reply by trifle** [reply]
That's indeed an important consideration. Tracking state across requests would require a lot of complexity though.

My personal intuition is to first implement caching for stateless requests to keep it simple. Anything on top should probably be handled by the client?

At least for my use case, I don't see why one would need to cache oauth flows (any cached traffic would be invalid anyways). But perhaps there are use cases where one wants to use mitm to replay/debug things like that. Although your comment above regarding mitmproxy is valid then, for debugging, that's probably the better tool.

Another source of complexity that I haven't thought about is simultaneous requests. As far as I remember, some browsers do use parallel requests to the same host; not sure how those would look like in mitm at the moment. If they create multiple simultaneous connections to a single host, that could lead to cache-related race conditions.

**Reply by synchronizing** [reply]
> My personal intuition is to first implement caching for stateless requests to keep it simple. Anything on top should probably be handled by the client?

This makes sense in the context of HTTP in specific, but I'm not sure how this system would fare out with different protocols. Regardless, I do agree with you - small steps first.

> Another source of complexity that I haven't thought about is simultaneous requests.

As far as I know simultaneous requests don't usually share any state variables - no? Frankly, my only knowledge of simultaneous requests is AJAX loading data from multiple endpoints to display on the page. Whatever the case, this ends up back in the "different protocol" conversation. 🤔

**Comment by synchronizing** [comment]
A possible solution to this would be to move more responsibility to the [`mitm.Protocol`](https://synchronizing.github.io/mitm/customizing/protocols.html) class and allow custom implementations of the actual `relay` functionality.

A bit of missing context: this project was built to be a support project for another project of mine whose intentions are to be a proxy gatherer, filter, and server. I specifically wrote `mitm` so that I have a method to read traffic between clients, proxies, and servers. The `mitm.Protocol` class is the way I envisioned myself and others to be able to build extensions for `mitm` - in my case, to hijack the client & server connection and embed as many proxies in the middle as needed until a valid response is returned.

In other words: the `mitm.Protocol` aspect of this project is/was the intended way to build projects like yours and mine. However and arguably, the name `protocol` doesn't translate to that and the current iteration of the `mitm.Protocol` does not allow as much customization as I would like for it to include. With your idea now in mind I'll see what I can do to update `mitm.Protocol` to allow something like caching and hijacking to work.

On the topic: any suggestions for a new name to replace `Protocol`?

**Reply by trifle** [reply]
Hm, I see the appeal of moving up into the protocol abstraction. That would allow protocols to choose whether to keep track of state or not. There are a number of generic names that come to mind, but I'm actually not sure protocol is misleading? Would you prefer "channel" or "layer" or "transport"?

By the way, your use case sounds interesting. Do you envision something like a proxy pool for crawling? That's actually something that would benefit from caching as well, since a cache can basically act as a "de-duplicator" for requests. IIRC there is one company offering something similar as a service. By the way, I guess you'd still need to parse responses to figure out which ones are successful, no? At least taking status codes into account, or even doing further validation of the response (i.e. if there are erros returned as json etc.). 

A last note - I'm sure you have had a look at other python proxy implementations. I did and took away a cautious lesson. For example, [python-proxy](https://github.com/qwj/python-proxy) is async as well and implements a host of protocols (including socks and ssh), but it has a ton of home-brew crypto implementations and a really dense code style. [Proxy.py](https://github.com/abhinavsingh/proxy.py) on the other hand is fairly well-structured but has a very complicated architecture including multiprocessing AND async executors in one (IIRC). That's why I think this project would really benefit from a small core while offloading complexity into custom middlewares and/or protocol implementations.

**Reply by synchronizing** [reply]
I have had a lot of conflicting thoughts for the naming since `middleware` is also being used. Maybe take inspiration from Django and have 'middlewares' and 'apps'? It does sounds weird, however, to say "`mitm` comes out of the box with HTTP(S) app"... not sure. 

One thing that I do want to maintain is the ability for the user to add multiple protocols/"apps" to `mitm`, but I'm unsure how that design would work with `relay` also being able to be modified. Will need to think a little bit more about this.

> Do you envision something like a proxy pool for crawling? That's actually something that would benefit from caching as well, since a cache can basically act as a "de-duplicator" for requests. 

Indeed - the idea would be to mass crawl free online proxies and put them through this filtering & serving system. Caching requests is something I did not consider for this use-case, but it is an interesting thought and something I will definitely look into as the prototype becomes more usable.

> By the way, I guess you'd still need to parse responses to figure out which ones are successful, no? At least taking status codes into account, or even doing further validation of the response (i.e. if there are errors returned as json etc.).

I wrote [`httpq`](https://github.com/synchronizing/httpq) for this specific case. It's also what's being internally used within `mitm` to identify if the protocol is HTTP/1.1. 

https://github.com/synchronizing/mitm/blob/9fabb564a7bb6f3e6da42ba325f3d797aec6c9e4/mitm/protocol.py#L127-L130

The idea I have in mind is to give users the ability to define site-specific HTTP response expectations - e.g. say "I expect Twitter to have these headers, and this JSON format." If it doesn't, then the project would re-route the request to a different proxy for a predefined `n` amount of times before sending back an error.

> A last note - I'm sure you have had a look at other python proxy implementations. I did and took away a cautious lesson. For example, [python-proxy](https://github.com/qwj/python-proxy) is async as well and implements a host of protocols (including socks and ssh), but it has a ton of home-brew crypto implementations and a really dense code style. [Proxy.py](https://github.com/abhinavsingh/proxy.py) on the other hand is fairly well-structured but has a very complicated architecture including multiprocessing AND async executors in one (IIRC).

Wow, those projects look really interesting and something I did not come across while doing research for exactly that. I haven't pushed the current code for the proxy project I spoke about, but internally, I'm using [`python-sockets`](https://github.com/romis2012/python-socks). I'm going to check out the projects you linked further and see if they prove better.

> That's why I think this project would really benefit from a small core while offloading complexity into custom middlewares and/or protocol implementations.

100% and my exact thoughts. I do think a few things should come out-of-the-box (like support for HTTP), but the vast majority of projects/ideas should be offloaded into their own repos. My intention with `mitm` is for others (and myself) to be able to `import mitm` and build on top it.

**Reply by trifle** [reply]
Thanks for the explanation - I think that the current implementation of protocols is a great idea.
It also seems that the architecture is geared towards a passive model: You're listening in on a connection and try to optimistically parse protocols, but with no guarantee that this works.

Considering what would need to happen to effectively cache http, I'm beginning to think that it might be necessary (or at least more straight-forward) to use a different model: To *terminate* the connection internally and initiate separate connections outwards. This would mean abandoning the passive model in order to, for example, re-assemble and perhaps decompress chunked requests or responses before relaying.

That said, I see the advantages of having a completely transparent and light-weight (i.e. passive) architecture. I'll think a bit about whether it's possible to reconcile both!

**Reply by synchronizing** [reply]
> It also seems that the architecture is geared towards a _**passive model**_: You're listening in on a connection and try to optimistically parse protocols, but with no guarantee that this works.
> 
> Considering what would need to happen to effectively cache http, I'm beginning to think that it might be necessary (or at least more straight-forward) to use a different model: _**To terminate the connection internally and initiate separate connections outwards.**_ 

Could you expand a little more on what you mean? I don't think I fully understand the bolded & italicized text.

**Reply by trifle** [reply]
Sure! First, it might be useful to explain some cases in which caching is actually useful, and also why I don't want to implement it client-side.

Consider a web crawl that is either distributed, i.e. using multiple clients and/or temporal, i.e. performs re-crawling of the same resources over time. In such a scenario, you might be using direct http requests or a full browser (i.e. selenium). All of these factors can lead to duplicate requests that put a lot of strain on the host and are suspicious, perhaps leading to blocking or rate limiting:

- there may be ubiquitous resources (such as stylesheets, JS, images etc.) that are requested by a browser. If you're restarting instances and clear caches, those get re-fetched frequently, especially when using multiple clients
- when clients are configured to start from a seed node, every client will begin from the same URLs and might re-fetch what others have already received. Caching would enable opportunistic cooperation where cached resources are "skipped" and clients can focus on extending the crawl
- resuming crawls usually either means tracking all state (can be very resource costly) or re-fetching everything. Working from a cache simplifies crawler logic by simply re-playing known content; in addition, a TTL can help selectively fetch expired resources
- errors, broken links and loops may lead to re-fetches of the same resource and might not be detectable ex ante
- re-crawling schedules can be too fast especially if crawling is decentralized with no central scheduler
- configuration and coding errors happen all too often when exploring new sites, thus it's easy to accidentally send off too many and identical requests

That's a high-level overview of why I'm so interested in caching. As you see, I'm coming from a very particular mindset of distributed work with soft coordination - i.e. an architecture where I assume clients are partly cooperative and distribute work, but they have no real-time synchronization (i.e. no distributed or centralized database, no postgres or cassandra).

Regarding mitm, my question is mainly two things:
1. If we follow the protocol implementation, at what point could we determine whether a connection can be duplicated?

If we're dealing with http, that's easy, since a GET is a single request. But with https, we have two steps with a CONNECT before. Both are already handled differently:
https://github.com/synchronizing/mitm/blob/9fabb564a7bb6f3e6da42ba325f3d797aec6c9e4/mitm/protocol.py#L195-L198

So it seems possible to just extend that logic and read far enough into a request to determine the target path (by the way, there are of course some nuances to watch out for when de-duplicating requests, such as ordering query parameters. The internet archive python ecosystem does this by using a custom URL representation called [SURT](https://github.com/rajbot/surt)).

It would then be possible to pre-empt the connection to return a cached version of the content. Currently, protocol.py establishes the relayed https connection *before* receiving the GET (if I read this correctly):
https://github.com/synchronizing/mitm/blob/9fabb564a7bb6f3e6da42ba325f3d797aec6c9e4/mitm/protocol.py#L187-L193

As soon as the GET parameters are known, I could look them up in any sort of storage (redis, pickle, sqlite ...) and return the full response.

2. There is the question of pooling connections to external hosts.

Ultimately, I think some sort of mechanism will be necessary to control how many connections and requests are made to external hosts / through external proxies. The simplest version is of course to leave that responsibility to clients. Next simplest would be to hold a single connection to each external host and serialize incoming requests. This is where my idea of connection termination came from: To make sure that multiple clients can efficiently use mitm.

(By the way, a very interesting scenario would be to have a cascade of caching instances of mitm on multiple machines.)

I hope this explains a bit better what I'm aiming at!

**Reply by synchronizing** [reply]
> That's a high-level overview of why I'm so interested in caching. As you see, I'm coming from a very particular mindset of distributed work with soft coordination - i.e. an architecture where I assume clients are partly cooperative and distribute work, but they have no real-time synchronization (i.e. no distributed or centralized database, no postgres or cassandra).

I see now, this makes complete sense.

> (1) If we follow the protocol implementation, at what point could we determine whether a connection can be duplicated?

I suppose the best way to go about this would be to handle the client & server connection separately all together. Instead of storing request/response pairs one would store the entire flow (from the first request to the last request). When and if the flow breaks out of the cached flow `mitm` replays the flow up to that point with the server and caches this undiscovered branch for the next client. If the flow ends in a branch that the `mitm` has seen it doesn't bother connecting with the server. `GET` requests are also intuitively the only safely cacheable type of requests, and even then, they are not safe from state synchronization - imagine the following three clients sending requests one after the other: 

1. 
    * Client to server (`GET`): What is my username?
    * Server to client: Y

2.
    * Client to server (`POST`): Set my username to 'X'
    * Server to client: Username set.
    * Client to server (`GET`): What is my username?
    * Server to client: X

3. 
    * Client to server (`GET`): What is my username?
    * Server to client: Y

With some internal modifications to how `Protocols` work (more added responsibilities) I think it would be possible to build a simple `HTTPCached(Protocol)` model, but the problems above will likely need to be documented for other users.

> (2) There is the question of pooling connections to external hosts.

This is something I've thought about as well but is something that could not be added to the `mitm`. Complications arise when stateful connections or protocols are being used. Two clear examples are:

* [TLS](https://stackoverflow.com/a/33681674/1305461) - after a successful `CONNECT` the sockets on the client & server side are now expecting only encrypted packets.
* SSH - after a successful login the protocol does not expect authentication on every packet.

For this to work my intuition is telling me it would only be possible with something like plain HTTP/1.1 requests, where the client/server connections are stateless. The work-around for this is to ensure that the clients are properly maintaining sessions, and not opening/closing connections with the same server when not needed.

**Comment by synchronizing** [comment]
Also - have you considered caching the request & responses on the client side? I very recently implemented a client-side caching system that uses [Redis](https://redis.io/) to cache outgoing requests to prevent 429's. Redis is thread-safe and gives you some decent flexibility when it comes to setting caching time. Perhaps something that could be of help.

**Comment by trifle** [comment]
(Sorry for the weird threading, I'm not sure whether replies are more readable than new messages).
I've read a bit about proxy.py's approach to caching (which shows really how complicated it is), and found that what I'm looking for is usually discussed under the label "VCR", which I didn't know before!

People seem to be using such caches to speed up tests mainly, but the overall architecture is not far from what I envisioned. Perhaps we'll find some interesting other projects as well for reference: https://github.com/search?l=Python&o=desc&q=vcr&s=stars&type=Repositories

[edit] And there's this: https://github.com/johtso/httpx-caching

**Reply by synchronizing** [reply]
I'll check out VCR this weekend and answer back on the larger reply you sent. Busy end to a busy sprint @ work!

**Reply by trifle** [reply]
Yeah, no pressure from my side! Best of luck with the sprint!

**Reply by trifle** [reply]
I have a few free minutes to try and understand how much work would be necessary for mitm to work on request/response pairs. Just to document what I'm thinking, I'll write some questions here.

One thing that I'm not quite sure I understand is how exactly connections look like under asyncio/concurrent conditions. At the start of a new connection, you're reading N bytes into the request to determine the protocol. This is to allow the additional "CONNECT" step in https.

If I read this correctly, you're relaying one request/response *pair* next:

https://github.com/synchronizing/mitm/blob/9fabb564a7bb6f3e6da42ba325f3d797aec6c9e4/mitm/mitm.py#L176-L180

Curiously, the order is first response and then request, but given that gather launches both concurrently, this should not matter for the practical order.

Both [_request](https://github.com/synchronizing/mitm/blob/9fabb564a7bb6f3e6da42ba325f3d797aec6c9e4/mitm/mitm.py#L103) methods (request and response) next concurrently read into whatever is being sent to them. They read data until there's no data left and then send the buffer downstream. However, there's also an `asyncio.event` which should mean that as soon as either end reaches the end, the other will also return eventually. The `at_eof` method is implemented by the underlying asyncio StreamReader IIRC.

It seems that it should be possible to serialize the request/response cycle. To do that, we'd need to serially await each step:

```python
await _relay(connection, event, Flow.SERVER_TO_CLIENT)
await _relay(connection, event, Flow.CLIENT_TO_SERVER)
```

And remove the synchronizing event from `_relay`. Doing that seems to work as far as I see here.

**Reply by trifle** [reply]
Ok, forcing serial handling works, it's just super slow. Perhaps there's a timeout in waiting for the line to be empty?
In any case, here's one full request/response cycle using the example from the README:

```
2022-02-13 16:50:03 INFO     MITM started on 127.0.0.1:8888.
2022-02-13 16:50:05 INFO     Client 127.0.0.1:52773 has connected.
2022-02-13 16:50:05 INFO     Client to server:

→ CONNECT httpbin.org:443 HTTP/1.0
→

2022-02-13 16:50:05 INFO     Connected to server 54.221.78.73:443.

<-> Relaying one ping pong # this is where the serialized handling starts
-> REQUEST TO SERVER
2022-02-13 16:50:05 INFO     Client to server:

GET /anything HTTP/1.1
Host: httpbin.org
User-Agent: python-requests/2.27.1
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive



<- RESPONSE TO CLIENT
2022-02-13 16:50:10 INFO     Server to client:

<response bytes>

2022-02-13 16:50:15 INFO     Server has disconnected.
2022-02-13 16:50:15 INFO     Client has disconnected.
```

**Reply by synchronizing** [reply]
The reason why we don't serialize the flow, e.g.

```python
await _relay(connection, event, Flow.SERVER_TO_CLIENT)
await _relay(connection, event, Flow.CLIENT_TO_SERVER)
```

is so that the client & server can have the freedom to communicate back-and-forth without control from `mitm`. In other words, if we serialize the flow (like the code above) we are effectively saying the server has to reply to the client before the client can reply back to the server, and once those two things happen, that's all the talking they get.

When the flow happens concurrently

https://github.com/synchronizing/mitm/blob/9fabb564a7bb6f3e6da42ba325f3d797aec6c9e4/mitm/mitm.py#L177-L180

What we are essentially doing is allowing the client and server to keep going back-and-forth until one of them has nothing else to say. Nothing else to say is not indicative by `eof` (end-of-file), but rather by there not being any more data being sent in a set amount of time by one side or the other - this happens via the `await asyncio.wait_for(..., self.timeout)`, which causes `event.set()` since `data = None` due to `TimeoutError`.

https://github.com/synchronizing/mitm/blob/9fabb564a7bb6f3e6da42ba325f3d797aec6c9e4/mitm/mitm.py#L115-L128

When that happens there are two things that can happen thereafter:

https://github.com/synchronizing/mitm/blob/9fabb564a7bb6f3e6da42ba325f3d797aec6c9e4/mitm/mitm.py#L167-L183

Either:

(a) we keep their connection alive until one side closes the connection (we get an `eof`)
(b) if `keep_alive` is set to `False` we just drop their connection on the `mitm` end.

The likely reason why serializing makes it slow is because you are not waiting from both the client & server at the same time

https://github.com/synchronizing/mitm/blob/9fabb564a7bb6f3e6da42ba325f3d797aec6c9e4/mitm/mitm.py#L118-L121

and instead you are waiting for one before the other - stopping the flow to wait on one before checking on the other. 

---

Serialization of the flow might work for vanilla HTTP since it's connection flow is always either:

* Client to server: `CONNECT` (`Protocol.HTTP`)
* Server to client: `OK` (`Protocol.HTTP`)
* [TLS Flow](https://www.cloudflare.com/learning/ssl/what-happens-in-a-tls-handshake/) (`Protocol.HTTP`)
* Client to server: Request (`MITM._relay`)
* Server to client: Response (`MITM._relay`)
* Either done, or restart the last two steps if `Keep-Connection: Alive`. (`MITM.mitm`)

or

* Client to server: Request (`MITM._relay`)
* Server to client: Response (`MITM._relay`)
* Either done, or restart the last two steps if `Keep-Connection: Alive`. (`MITM.mitm`)

but it likely would not work for protocols like FTP, SSH, or even protocols that leverage HTTP, like [websockets](https://datatracker.ietf.org/doc/html/rfc6455#section-1.2), which could keep getting looped in `_relay` and `MITM.mitm`.

---

As it relates to VCR it seems like it's also client-sided, which I think raises the question on whether or not it can be used with distributed clients with soft coordination. If in your case coordination can be done on the client-side code, I still believe Redis is the best way to go about it. The implementation is actually quite simple - you create a class `Client` that wraps `request`/`httpx`/etc., and has the method `.get(url)` to retrieve URLs - in that method you do:

1. Attempt to get Redis cache using URL as key.
   1. If cache exists, return to the client.
   2. If cache does not exist, send request.
      1. Set Redis cache with key as the URL and response bytes as value. Set the expiration time for the key.

As long as all the clients are connected to a centralized Redis service the cache usage should be fluent. This is a very basic implementation of something I built internally for my workplace (we scrape millions of datapoints from different APIs through hundreds of simultaneously running "scraper" nodes), and weren't sure if we are using `requests` or `httpx` long-term.

**Comment by trifle** [comment]
Hi, just a short update on this!
I followed your advice above and had another look at implementing caching using mitmproxy.
Interestingly enough, there has been lots of work in master to support a (more or less) pure async mode now.

Using that as a base, I've built a POC that works pretty well. There are some interesting rough edges, of course: Re-assembling the request/response pairs stalls the connection for large downloads, memory balloons and some clients time out. But it's a solid place to start from. I've even had some success caching POSTs (using host,query and a hash of the posted data as the caching key).

All in all, I'd say that my use case probably works better on a request/response level than on a connection level.

**Reply by synchronizing** [reply]
My apologies @trifle, I must have not seen these replies when they came in. Hope your project has been going well!
